rust/crd/src/authentication.rs

```diff
@@ -1,95 +1,318 @@
+        )
+    }
+
+    /// Return the (first) TLS `AuthenticationClass` if available
+    pub fn get_tls_authentication_class(&self) -> Option<&AuthenticationClass> {
+        self.resolved_authentication_classes
+            .iter()
+            .find(|auth| matches!(auth.spec.provider, AuthenticationClassProvider::Tls(_)))
+    }
+
+    /// Validates the resolved AuthenticationClasses.
+    /// Currently errors out if:
+    /// - More than one AuthenticationClass was provided
+    /// - AuthenticationClass provider was not supported
+    pub fn validate(self, server_and_internal_secret_class: Option<String>) -> Result<Self, Error> {
+        if self.resolved_authentication_classes.len() > 1 {
+            return Err(Error::MultipleAuthenticationClassesProvided);
+        }
+
+        for auth_class in &self.resolved_authentication_classes {
+            match &auth_class.spec.provider {
+                AuthenticationClassProvider::Tls(_) => {}
                 _ => {
-                    return Err(Error::AuthenticationClassProviderNotSupported {
-                        authentication_class_provider: authentication_class
-                            .spec
-                            .provider
-                            .to_string(),
-                        authentication_class: ObjectRef::<AuthenticationClass>::new(
-                            &druid_tls.authentication_class,
-                        ),
+                    return Err(Error::AuthenticationProviderNotSupported {
+                        authentication_class: ObjectRef::from_obj(auth_class),
+                        provider: auth_class.spec.provider.to_string(),
                     })
                 }
             }
         }
 
-        Ok(druid_authentication_config)
+        if let Some(tls_auth_class) = self.get_tls_authentication_class() {
+            match &server_and_internal_secret_class {
+                Some(server_and_internal_secret_class) => {
+                    // Check that the tls AuthenticationClass uses the same SecretClass as the Druid server itself
+                    match &tls_auth_class.spec.provider {
+                        AuthenticationClassProvider::Tls(tls) => {
+                            if let Some(auth_class_secret_class) = &tls.client_cert_secret_class {
+                                if auth_class_secret_class != server_and_internal_secret_class {
+                                    return Err(Error::TlsAuthenticationClassSecretClassDiffersFromDruidServerTls { authentication_class:  ObjectRef::from_obj(tls_auth_class), server_and_internal_secret_class: server_and_internal_secret_class.clone() });
+                                }
+                            }
+                        }
+                        _ => unreachable!(
+                            "We know for sure we can only have tls AuthenticationClasses here"
+                        ),
```

Comment:
That is 100% sure? Otherwise if there is any possibility, i would go for an error (to show up as event).

Reply:
120% sure :)